### ISSUE TITLE
20200701 북스터디

### ISSUE THREAD

**Issue by sangyun-han** [sangyun-han/study#3]
### Data flow graph
데이터 처리 operator 사이로 데이터가 어떻게 흐르는지를 기술한 그래프(directed graph)
- node == operator
- edge는 의존 관계를 표현

**Comment by sangyun-han** [comment]
## 윈도우 연산

### Tumbling window
고정 길이로 겹치지 않도록 버킷을 할당하여 처리하는 방식으로 count-based와 time-based 텀블링 윈도우로 나뉜다.
- count-based는 인입된 데이터의 갯수를 기준으로 윈도우를 할당하여 연산을 수행
- time-based는 데이터가 들어오는 단위 시간을 기준으로 윈도우를 할당하여 연산을 수행

### Sliding window
앞뒤 버킷이 어느 정도 중첩되도록 할당하여 처리하는 방식으로 기존 네트워크 방식에서 많이 사용한다.

### Session window
텀블링이나 슬라이딩 윈도우로 적용할 수 없는 현실 세계의 시나리오에 유용하다.
갯수, 시간 기준이 매번 바뀌는 경우, 세션이 종료됐다고 판단할 수 있는 비활동 시간을 session gap이라고 정의하여 이벤트들을 세션으로 그룹핑한다.




## 시간 시멘틱

### processing time
스트림 처리 연산자가 실행 중인 장비의 시계에서 측정한 시간을 처리 시간이라고 한다. [관찰자 입장에서의 시간]

### event time
이벤트가 실제 발생한 시간으로 이벤트 안에 포함된 타임스탬프를 기반으로 한다. [주체(생산자) 입장에서의 시간]
어떠한 이유에서든 지연되서 전달된 이벤트라도 이벤트 시간을 기준으로 하면 발생했던 시간과 순서를 제대로 반영하여 나타낸다.

### ingestion time
특정 이벤트가 flink 내의 특정 스트림에 인입된 시간. 
특정 이벤트가 플링크에 유입된 후에 할당하는 타임스탬프에 의해 결정된다.

**Comment by sangyun-han** [comment]
## FlinkML
머신러닝 개발자들이 최소한의 glue code만 작성해도 문제가 없도록 구현돼 있다.
FlinkML은 파이썬의 scikit-learn과 Spark의 MLlib에서 아이디어를 얻었다.

- data transformation & 학습용 파이프라인
- pre-processing : 
  - feature scaling
  - 다항 피처 기반 메퍼 (polynomial feature based mapper)
  - 피처 해싱(feature hashing)
  - 텍스트 데이터의 피처 추출
  - 차원 축소화 기법 (dimensionality reduction)
- 모델 셀렉션, 성능 평가 : 
  - 모델 선택과 평가를 위한 교차 검증 기법
  - 하이퍼 파리미터 최적화 기법

### 지도 학습
- 최적화 프레임워크
- SGD(Stochastic Gradient Descent) 기법
- L-BFGS 기법
- GLM(Generalized Linear Models) 기법
- 다중 선형 회귀 모델 기법
- LASSO, Ridge 회기 기법
- 다중 클래스 로지스틱 회귀 기법
- 랜덤 포레스트 알고리즘
- SVM
- 의사결정 트리 알고리즘

### 비지도 학습
- 클러스터링
- k-means 클러스터링
- PCA(Principal Component Analysis)

### 추천
- ALS(Alternating Least Squares)

### 텍스트 분석
- LDA(Latent Dirichlet Allocation)

**Comment by sangyun-han** [comment]
## 지원하는 커넥터

- Apache Kafka (source/sink)
- Apache Cassandra (sink)
- Amazon Kinesis Streams (source/sink)
- Elasticsearch (sink)
- Hadoop FileSystem (sink)
- RabbitMQ (source/sink)
- Apache NiFi (source/sink)
- Twitter Streaming API (source)
- Google PubSub (source/sink)


- Apache ActiveMQ (source/sink)
- Apache Flume (sink)
- Redis (sink)
- Akka (sink)
- Netty (source)
